### ISSUE TITLE
Torch compiled FLCE is 2x faster than the current FLCE

### ISSUE THREAD

**Issue by ByronHsu** [linkedin/Liger-Kernel#227]
### 🚀 The feature, motivation and pitch

We can leverage torch compile to fuse the things we cannot fuse now like upcasting, contiguous call, etc

![image](https://github.com/user-attachments/assets/94b377d3-6530-43f5-9f88-1c7108a4746f)

Sample code: https://gist.github.com/Chillee/22cd93e11b887db1f596ab754d60a899#file-lce_benchmark-py
Provided by the brilliant @Chillee

**Comment by wizyoung** [comment]
Actually, if we align the CHUNK_SIZE of the Torch-compiled FLCE with the strategy used in Liger's FLCE, the compiled version is only slightly faster than the Liger version, but it does require a bit more memory as well. The advantage of the Torch-compiled version is its flexibility; implementing the Gemma2 softcap logits is very straightforward, whereas I struggled for some time to achieve consistent accuracy with this in Liger.

**Comment by wizyoung** [comment]
By setting CHUNK_SIZE and run benchmark script:
![image](https://github.com/user-attachments/assets/2de14072-7a48-4439-92d9-d460c0af2f0b)
![image](https://github.com/user-attachments/assets/8bc4d1c3-0fb6-489b-a4b6-4706f8fb2e76)

**Comment by Chillee** [comment]
@wizyoung how are you setting the chunk size? I wasn't able to get the liger kernel to perform much better even when changing the chunk size.

**Comment by wizyoung** [comment]
@Chillee By referencing https://github.com/linkedin/Liger-Kernel/blob/main/src/liger_kernel/ops/fused_linear_cross_entropy.py#L23. I mean, chaning chunk size in torch compiled FLCE. Your default chunk size is 1024, and I change to 256. Then I have:
![image](https://github.com/user-attachments/assets/06e43de9-bd5f-4d08-a00e-cef45fa1f80f)
By only keep benchmark test with liger and compiled chunkce:
![image](https://github.com/user-attachments/assets/660c9584-c1b7-4eb8-9e8b-3ed517b8a20b)
![image](https://github.com/user-attachments/assets/ce8ef2b7-442c-47c3-9c17-5e2e203939f0)

**Comment by wizyoung** [comment]
I have done some quick tests with different B, T, D and V to mimic my training conditions(llama3 and gemma2) in my env, my conclusion is that torch compiled flce is indeed faster, but has worse memory management.

**Comment by wizyoung** [comment]
https://gist.github.com/wizyoung/5330ad501e73a97dfe2f0088decdb1ca
I have implemented a version of torch.compile chunked_lce that supports soft caps and passes all numerical accuracy tests in benchmark_fused_linear_cross_entropy.py modified from this repo. My main concern is the frequent changes in input shape, which result in varying chunk sizes. To mitigate this overhead, I used `torch.compile(dynamic=True, options={"shape_padding": True})`. However, I am still uncertain about its effectiveness and look into it during actual training.

**Comment by Chillee** [comment]
@wizyoung I agree there's some additional memory overhead (in particular, I think we don't inplace the addmm), but the additional memory is generally pretty negligible here, no?

For example, if I change the chunk size from 256 to 512, torch.compile performance improves from 186ms down to 153, while memory only increases from 1.48 GB to 1.54 GB.

If I try increasing the chunk size of Liger, it doesn't seem to increase the performance as much as the torch.compile version

**Comment by ekojsalim** [comment]
Curious how this compares with [JonasGeiping/linear_cross_entropy_loss ](https://github.com/JonasGeiping/linear_cross_entropy_loss), but `torch.compile` seems good enough though.

**Comment by Chillee** [comment]
@ekojsalim In my brief testing, it seems like it's both faster and uses less memory.

**Comment by wizyoung** [comment]
> @wizyoung I agree there's some additional memory overhead (in particular, I think we don't inplace the addmm), but the additional memory is generally pretty negligible here, no?
> 
> For example, if I change the chunk size from 256 to 512, torch.compile performance improves from 186ms down to 153, while memory only increases from 1.48 GB to 1.54 GB.
> 
> If I try increasing the chunk size of Liger, it doesn't seem to increase the performance as much as the torch.compile version

Yes, the increase in memory usage is generally negligible. My primary concern is the running time overhead, specifically that the B*T varies significantly and is not a multiple of the chunk size, leading to frequent calls of recompile(I add `TORCH_LOGS="recompiles"` to find that). Therefore, I use `torch.compile(dynamic=True, options={"shape_padding": True})` as documented; however, I am uncertain about its actual effectiveness.
I did an expensive benchmark test using the script of this repo by setting `BT = [2**12] + (np.random.randint(2**12, 2**15 + 1, 80) + np.random.randint(0, 1001, 80)).tolist() + [2**15]` and keeping H=4096 and V=128256. And chunk_size is 1024.
![image](https://github.com/user-attachments/assets/37da2864-a44a-47dc-ab05-2ce7e44550c7)
![image](https://github.com/user-attachments/assets/82cbaf19-0657-4810-9f48-acf98de3c4fe)

**Comment by Chillee** [comment]
@wizyoung Can you post your actual benchmark?

**Comment by wizyoung** [comment]
@Chillee I have updated my scripts here: https://gist.github.com/wizyoung/5330ad501e73a97dfe2f0088decdb1ca
